### ISSUE TITLE
The typing of dataclass field's default_factory replaces explicit typing in strict mode

### ISSUE THREAD

**Issue by JordonPhillips** [microsoft/pyright#10277]
The typing of `field`'s `default_factory` seems to be replacing explicit typing on the property in strict mode.

```python
from dataclasses import dataclass, field
from typing import Mapping


@dataclass
class Example:
    prop: Mapping[str, str] = field(default_factory=dict)
```

```
[tool.pyright]
typeCheckingMode = "strict"
```

```
/tmp/repro.py
  /tmp/repro.py:7:5 - error: Type of "prop" is partially unknown
    Type of "prop" is "dict[Unknown, Unknown]" (reportUnknownVariableType)
1 error, 0 warnings, 0 informations
```

This is reproducible both in command line and in VSCode with version `1.1.398`. This seems to have been introduced by #10217. While I agree in the general sense that Unknowns should result in pyright errors in strict mode, I don't agree that it should apply in this case. It's the equivalent of the following snippet that passes:

```python
from typing import Mapping


class NonDataclassExample:
    property: Mapping[str, str]

    def __init__(self, prop: Mapping[str, str] | None = None) -> None:
        self.prop = prop if prop is None else dict()
```

**Comment by erictraut** [comment]
Yes, this is intended behavior. 

This behavior is due to "type narrowing upon assignment". When assigning a value, pyright validates that the RHS type is "assignable to" the declared type of the LHS. It then applies type narrowing if the RHS type is narrower than the declared type. For example:

```python
x: float = 1
reveal_type(x) # Literal[1]

y: Mapping[str, str] = {}
reveal_type(y) # dict[str, str]
```

Type narrowing doesn't "replace" the declared type, but it does update the narrowed type of the target.

This occurs regardless of whether you're in strict mode.

The type narrowing logic prior to 1.1.398 was inconsistent and had some bugs. The change in behavior that you're seeing here was intentional.

If you want to eliminate the strict-mode error in this case, you could do the following:

```python
@dataclass
class Example:
    prop: Mapping[str, str] = field(default_factory=dict[str, str])
```

**Comment by travisdowns** [comment]
For what it's worth, using the fully qualified type on both sides as recommended a couple of comments up is flagged as a straight up high-confidence bug by code review tools such as GH copilot reviewer (though I suppose this can change with time).

**Comment by travisdowns** [comment]
> Having to write and maintain this is a punishment:

> `input_queue: Queue[tuple[str, Log, Queue[Any]]] = field(default_factory=Queue[tuple[str, Log, Queue[Any]]])`

For what it's worth, in this case you don't need to repeat the type on the LHS and RHS, you can omit the type hint entirely and rely on the return type of `field`:

```
input_queue = field(default_factory=Queue[tuple[str, Log, Queue[Any]]])`
```

That doesn't work for the case in the OP where the LHS was a `Mapping` and the field factory returned a `dict` but it works for many other common cases to avoid the duplication.

**Comment by Jackenmen** [comment]
> For what it's worth, in this case you don't need to repeat the type on the LHS and RHS, you can omit the type hint entirely and rely on the return type of `field`:

`dataclasses` require type annotations for their fields, so you actually can't omit it:
https://github.com/python/cpython/blob/e07b78c01a57c5921614388e287c584d19a3e40c/Lib/dataclasses.py#L1035-L1038

**Comment by travisdowns** [comment]
You are right, the code i posted type checks but fails at runtime!

**Comment by rax-V** [comment]
If you get annoyed by this enough like I did...
```python
import dataclasses
import typing


class _DefaultFactory:
    """Descriptor that infers default_factory from field's type annotation."""

    def __init__(self, **kwargs: typing.Any) -> None:
        self._kwargs = kwargs

    def __set_name__(self, owner: type, name: str) -> None:
        hint = typing.get_type_hints(owner).get(name)
        if not callable(hint):
            raise TypeError(f"default_factory: {name} needs callable type, got {hint}")
        setattr(owner, name, dataclasses.field(default_factory=hint, **self._kwargs))


def default_factory(**kwargs: typing.Any) -> typing.Any:
    """Dataclass field with default_factory inferred from type annotation.

    Usage:
        items: list[str] = default_factory()
        private: list[bytes] = default_factory(init=False, repr=False)
    """
    return _DefaultFactory(**kwargs)
```

**Comment by Kludex** [comment]
> This breaking change is way too punishing to dataclass users. The code in the original issue is aligned with the dataclasses pep. It uses `list` and `dict` constructors without type arguments:
> 
> [peps.python.org/pep-0557#a-complicated-example](https://peps.python.org/pep-0557/#a-complicated-example)
> 
> Adding the extra type arguments is a chore and introduces line noise. Our medium-sized project has around 100 lines with `field(default_factory=...)`, where some are of this length:
> 
> ```
>    input_queue: Queue[tuple[str, Log, Queue[Any]]] = field(default_factory=Queue)
> ```
> 
> Having to write and maintain this is a punishment:
> 
> ```
>    input_queue: Queue[tuple[str, Log, Queue[Any]]] = field(default_factory=Queue[tuple[str, Log, Queue[Any]]])
> ```
> 
> Why not support the naked constructor? The eta-expanded version still type-checks, after all:
> 
> ```
>    input_queue: Queue[tuple[str, Log, Queue[Any]]] = field(default_factory=lambda: Queue())
> ```
> 
> The downside of the lambda, beyond the extra line noise, is that the repr for the field is less informative.
> 
> Please reconsider this breaking change.

I have to agree with this.

---

Can we have a specific error for this behavior then?

**Comment by technillogue** [comment]
this is kind of ridiculous as "intended behavior":

```
@dataclass
class Spam:
    eggs: list[str] = field(default_factory=list) # reportAssignmentType
    ham: list[str] = field(default_factory=lambda: list()) # ok
    bacon: list[str] = field(default_factory=list[str]) # "correct"
```

ty, mypy, and pyre are all completely okay with that code. please add an error type for this behavior that can be disabled
